packages/react-native/ReactCommon/react/renderer/graphics/platform/ios/react/renderer/graphics/HostPlatformColor.h

```diff
@@ -8,32 +8,108 @@
 #pragma once
 
 #include <react/renderer/graphics/ColorComponents.h>
+#include <react/utils/hash_combine.h>
 #include <cmath>
 
 namespace facebook::react {
 
-using Color = int32_t;
+enum ColorKind { UndefinedKind, ComponentsKind, DynamicKind };
+
+struct DynamicColor {
+  int32_t lightColor = 0;
+  int32_t darkColor = 0;
+  int32_t highContrastLightColor = 0;
+  int32_t highContrastDarkColor = 0;
+
+  bool operator==(const DynamicColor& other) const {
+    return lightColor == other.lightColor && darkColor == other.darkColor &&
+        highContrastLightColor == other.highContrastLightColor &&
+        highContrastDarkColor == other.highContrastDarkColor;
+  }
+};
+
+class Color {
+ public:
+  Color() : Color(UndefinedKind) {}
+  Color(DynamicColor dynamicColor)
+      : kind_(DynamicKind),
+        dynamicColor_(dynamicColor),
+        color_(dynamicColor.lightColor) {}
+  Color(ColorComponents components) {
+    float ratio = 255;
+    kind_ = ComponentsKind;
+    color_ = ((int)round(components.alpha * ratio) & 0xff) << 24 |
+        ((int)round(components.red * ratio) & 0xff) << 16 |
+        ((int)round(components.green * ratio) & 0xff) << 8 |
+        ((int)round(components.blue * ratio) & 0xff);
+  }
+  int32_t getColor() const {
+    return color_;
+  }
+  ColorKind getKind() const {
+    return kind_;
+  }
+  DynamicColor getDynamicColor() const {
+    return dynamicColor_;
+  }
+  bool operator==(const Color& other) const {
+    return kind_ == other.kind_ && color_ == other.color_ &&
+        dynamicColor_ == other.dynamicColor_;
+  }
+  bool operator!=(const Color& other) const {
+    return kind_ != other.kind_ || color_ != other.color_ ||
+        dynamicColor_ != other.dynamicColor_;
+  }
+
+ private:
+  Color(ColorKind kind) : kind_(kind), color_(0) {}
+  ColorKind kind_;
+  int32_t color_ = 0;
+  DynamicColor dynamicColor_;
```

Comment:
can we use [`std::variant<int32_t, DynamicColor>`](https://en.cppreference.com/w/cpp/utility/variant) instead?
This optimize the space allocated (using the max between `int32_t` and `DynamicColor` instead of using both). It also provide the similar apis as the `getXXX` you implemented.